product editor 구현

altmit opened the pull request:
## Description
- 아이템 추가/수정 가능한 editor 구현
- useInput hook 추가

## Key changes

https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/41321198/d8434055-f29e-4238-9b3e-584a09e5b00f


- product editor는 panel로 구현 했습니다.
  - 해당 panel은 layout에 조건부 랜더링 됩니다.
  - panel의 상태는 useProductEditorStore.ts에서 관리할 수 있습니다.
  - useProductEditorStore만 가져와서 사용하면 어느 페이지에서도 사용 가능합니다.
- useInput hook 구현 했습니다.
  - state와 onChange를 모두 관리하는데 반복적인 js가 너무 많아서 추가 했습니다.
  - login, signup 같은 input 사용처에 추가 적용하면 좋을 것 같습니다. 
-  ts compilerOptions이 NS2020으로 되어 있어 replaceAll을 못하는 현상이 있어 NSNext로 변경 했습니다.

## To reviewers
- 리팩토링이 필요한 부분이 조금 있습니다.
  - type에 카테고리 관련 중복적인 type이 조금 있는데 리팩토링 작업을 한번에 묶어서 진행하면 좋을 것 같습니다.
- 상세 페이지가 구현 된다면 추가, 수정이 끝나면 해당 아이템 상세 페이지로 이동하는 로직 추가가 필요 합니다.
- 지금 editor에서 이미지를 업로드하고 input에 타이핑을 하면 이미지가 재랜더링 되며 깜빡이는 이슈가 있는데 추후에 수정할 예정입니다.
## Issue
- #83

altmit committed:
feat #82 : 추천 카테고리 관련 api, msw 구현

altmit committed:
chore #82 : ts compilerOptions ESNext로 변경

altmit committed:
feat #82 : useInput 구현

altmit committed:
chore #82 : lodash 라이브러리 추가

altmit committed:
refactor #82 : useItemQuery 변경

- useResetRecommendCategory 구현
- query key 상수화

altmit committed:
refactor #82 : category 관련 타입 선언

- 중복적으로 비슷한 타입이 있어서 리팩토링 필요

altmit committed:
feat #82 : 상품 수정, 추가 api 추가 및 msw 구현

altmit committed:
feat #82 : fab 버튼 구현 및 상품 등록 panel 연결

altmit committed:
feat #82 : 상품 추가, 수정 panel을 관리할 zustand store 추가

altmit committed:
feat #82 : 상품 추가, 수정을 위한 panel 구현

altmit committed:
refactor #82 : 불필요한 라이브러리 삭제

altmit committed:
refactor #82 : 불필요한 console.log 삭제

silvertae commented on src/api/queries/useItemQuery.ts:
```diff
@@ -50,7 +57,22 @@ export const useGetFavorites = (categoryId?: number) => {
+  return () =>
+    queryClient.resetQueries({
+      queryKey: [RECOMMEND_CATEGORY],
+      exact: false,
```
queryKey 배열 내부에 별도의 Array key가 없으면 `exact: true`로 해도 잘 동작하지 않을까요?

altmit replied:
이 부분 기존 코드를 복붙하고 exact를 지운다는거 까먹고 커밋했네요😅

silvertae commented on src/hooks/useInput.ts:
input을 커스텀 훅으로 따로 뺀 거 좋네요 👍

silvertae commented on src/mocks/itemsHandlers.ts:
```diff
@@ -38,6 +38,24 @@ export const itemsHandlers = [
+
+    return res(ctx.status(200), ctx.json({ categories: result }));
+  }),
+  rest.post(API_ENDPOINT.ITEMS, (_, res, ctx) => {
+    return res(ctx.status(201), ctx.json({ itemId: 3 }));
+  }),
+  rest.put(`${API_ENDPOINT.ITEMS}/:id`, (_, res, ctx) => {
+    return res(ctx.status(201), ctx.json({ itemId: 3 }));
+  }),
```
명세서에 나와있는 status code 생각안하고 그냥 200으로 다 써놨었는데 저도 찾아서 수정해야겠네요. 😅

silvertae commented on src/mocks/itemsHandlers.ts:
```diff
@@ -38,6 +38,24 @@ export const itemsHandlers = [
+    const clonedArr = [...categoryData.categories];
+    const result = [];
+
+    for (let i = 0; i < 3; i++) {
+      const randomIndex = Math.floor(Math.random() * clonedArr.length);
+      const [item] = clonedArr.splice(randomIndex, 1);
+      result.push(item);
+    }
```
mock data라 상관없을 것 같긴 한데 이렇게 하면 result 배열에 중복된 item이 들어가진 않나요?
그리고 아래 categoryData를 사용해서 만드는 거라 불필요한 iconName도 같이 들어가긴 하겠네요 ㅋㅋ

altmit replied:
splice를 하고 있기 때문에 중복된 값은 들어가지 않습니다.
불필요한 iconName은 mock data이기도 하고 받는 부분에서 type을 지정해 놓아서 크게 상관 없을 것 같네요

silvertae commented on tsconfig.json:
```diff
@@ -1,8 +1,8 @@
 {
   "compilerOptions": {
-    "target": "ES2020",
+    "target": "ESNext",
```
ES2020 일 때는 replaceAll 메서드를 사용할 수 없나요? 신기하네요.

altmit replied:
2021부터 가능하다고 하는 것 같아요

silvertae commented on src/components/Layout.tsx:
```diff
@@ -1,14 +1,19 @@
     <Wrapper $width={screenWidth} $height={screenHeight}>
       <Outlet />
       <Footer />
+      {isOpenEditor && <ProductEditorPanel />}
```
상품 등록/수정 패널을 내부에서 렌더링 되도록 한 이유가 궁금합니다.
홈, 판매내역, 관심목록 페이지에서 볼 수 있는 상품 상세페이지에서 바로 수정 페이지를 열 수 있게 하기위함 인 건가요?

altmit replied:
라우팅이 아닌 패널을 사용한 이유를 물어보는 질문이 맞을까요?

맞다면 2가지 이유가 있습니다.

1. 우선 당근도 저희 패널처럼 페이지 전환이 아닌 모달과 같은 레이어를 하나 더 띄우는 방식을 사용합니다.
2.  보안상의 이슈도 있을 것 이라 판단 했습니다.
    - 추가의 경우 상관 없지만 수정의 경우 라우팅 url에 아이템의 id가 노출됩니다.
    - 이떄 사용자가 이 id를 임의로 특정 주소로 움직일 수 있다는 뜻이기도 하죠
    - ex) `/editor/2`, `/editor/3` 같이 url을 변경하는 것 만으로도 특정 아이템의 수정 페이지를 접속할 수 있습니다.
    - 물론 서버 특에서 막을 수 있겠지만 시도 조차 못하게 하는게 좋다 판단 했습니다.

silvertae replied:
아 Layout에 ProductEditorPanel을 두신 이유를 여쭤봤던 건데 질문을 잘 못 적은것 같네요. 😅
그래도 답변 감사합니다! 2번째 이유는 생각못했던 장점이었는데 아주 좋은 것 같네요.

+) 저도 확인해봤는데 당근마켓은 상품 등록 화면은 확실히 아래에서 위로 패널로 나타나듯이 나오는 것 같더라구요.
상품 상세는 옆으로 넘어가듯이 나와서 이게 패널인지 페이지인지 아리까리 하네요.

그리고 테스트해보니 상품 등록을 마치면 기존에 열려있던 상품 등록 패널이 아래로 닫히고 홈이 보였다가 잠시 후에 옆으로 넘어가면서 상품 상세페이지가 나오는 것처럼 보였습니다.

silvertae commented on src/stores/useProductEditorStore.ts:
```diff
@@ -0,0 +1,48 @@
+  productId: undefined,
+  productData: undefined,
+  openPanel: params => {
+    if (params.mode === 'edit') {
+      set({
+        isOpen: true,
+        productData: params.data,
+        productId: params.id,
+        editorMode: params.mode,
+      });
```
이 부분을 보면 openPanel을 실행할 때 전달하는 param으로 productData를 store에 set 해주고 있는데요.

위에 타입을 보니 productData는 상품 수정 페이지를 GET 해오는 요청에 대한 응답 데이터인 것 같은데 이 데이터를 요청하는 과정은 openPanel을 호출하기 이전에 일어나는건가요?

ProductEditorPanel 내부에서 `useGetProductEditorData` 이런 식의 useQuery 훅을 두는 것으로 생각했는데, 패널을 여는 순간에 이미 서버에서 받아온 data를 set 해주어야 하네요.

altmit replied:
음.. 지금 로직이라면 openPanel을 하기 직전에 이미 productData를 가지고 있어야 하긴 하겠네요.

수정한다면 store에서 api를 호출할 수 있지 않을까 싶기도 하네요.

ex)
```ts
openPanel: async params => {
  if (params.mode === 'edit' && params.id) {
      const fetchedProductData = await useGetProductEditorData(params.id);
      set({
        isOpen: true,
        productData: fetchedProductData,
        productId: params.id,
        editorMode: params.mode,
      });
    }
  } else if (params.mode === 'add') {
    set({ isOpen: true, productData: undefined, editorMode: params.mode });
  }
```
가능한지 여부는 수정 해보면서 확인 필요할 것 같네요

silvertae commented on src/page/productEditor/LocationDropdown.tsx:
```diff
@@ -0,0 +1,60 @@
+            </MenuItem>
+          );
+        })}
+        <MenuItem onAction={openModal}>내 동네 설정하기</MenuItem>
+      </Dropdown>
+      {isModalOpen && (
+        <HomeLocationModal isOpen={isModalOpen} onClose={closeModal} />
+      )}
```
Home에서 사용하던 동네 설정 모달을 등록 페이지에서도 열어볼 수 있게 하셨군요.
전 기획서에 따로 언급이 없길래 그냥 이미 설정된 유저 동네 1~2개만 드롭다운으로 보여주고 그 안에서만 바꿀 수 있게 하는걸로 생각했네요 🤔

altmit replied:
넵 처음에는 동네 추가/삭제도 가능하게 하려 했는데 버그가 있는 부분 수정한는걸 잊고 커밋 했네요.

그래서 아마 `내 동네 설정`은 지우고 새로 커밋 할 것 같네요

silvertae commented on src/page/productEditor/ProductContent.tsx:
```diff
@@ -0,0 +1,37 @@
+export function ProductContent(props: ProductContentProps) {
+  const textareaRef = useRef<HTMLTextAreaElement>(null);
+
+  const onInput = () => {
+    if (textareaRef.current) {
+      textareaRef.current.style.height = 'auto';
+      textareaRef.current.style.height = `${textareaRef.current.scrollHeight}px`;
+    }
+  };
```
element.style 로 height를 직접 지정하면 안 될 것 같네요. 높이가 계속 늘어나면서 화면밖으로 나가고 있습니다. 😨

![스크린샷 2023-09-14 오후 4 01 09](https://github.com/max2023-4th-project-01/FE-Cokkiri-Market/assets/76121068/5ce3b658-f925-4000-965d-2cbeceb46c53)

altmit replied:
Body에 스크롤을 추가 안 했군요 😅
추가 하겠습니다

silvertae commented on src/page/productEditor/ProductEditorPanel.tsx:
```diff
@@ -0,0 +1,457 @@
+    setSelectedCategoryId(id);
+  };
+
+  const selectCategory = (id: number, name: string) => {
```
테스트 해봤을 때 사용자가 카테고리를 세 개 중 하나를 선택해도 다른 문자를 입력하면 기존에 선택했던 카테고리가 초기화 되고 3개 모두 바뀌네요. 다른 두 개는 바뀌더라도 이미 선택된 하나의 카테고리는 유지되어야 할 것 같습니다.

+) 현재는 add 모드에서 처음 랜덤한 카테고리 3개가 나타날때 무조건 첫번째 카테고리가 선택되어서 나오네요. 의도하신 동작이 맞을까요?

altmit replied:
둘 다 의도한 동작이긴 합니다.

- 구현 당시에 타이틀이 입력이 달라질 때 마다 새로운 카테고리를 3개를 보여주는게 맞다 판단 했습니다.
  - 유저 흐름을 에를 들면, `나이키 운동화`를 입력 했을 경우 `신발, 의류, 남성의류` 등 같이 나왔다가 `삼성 모니터로` 바꾼다면 `디스플레이, 가전, 디지털기기`로 바뀌는 느낌을 생각 했습니다.
- 만약 정말 title에 맞는 추천 카테고리 3개가 온다면 가장 적합한 카테고리가 첫번째 오겠다는 생각이 들어 첫번째 카테고리를 선택하게 했습니다.
  - , `나이키 운동화`를 입력 했을 경우 `신발, 의류, 남성의류`  중 신발
  - `삼성 모니터로` 를 입력 했을 경우 `디스플레이, 가전, 디지털기기` 중 디스플레이

silvertae commented on src/page/productEditor/ProductEditorPanel.tsx:
```diff
@@ -0,0 +1,457 @@
+  const isEdit = editorMode === 'edit';
+  const isError = isEdit && (!productData || !productId);
+
+  const selectLocationId = (productId: number) => {
+    return userLocationData!.locations.map(location => ({
+      ...location,
+      isSelected: location.id === productId,
+    }));
+  };
```
함수 이름이 selectLocationId라 뭔가 지역 아이디를 선택하는 동작을 할 줄 알았는데, return 값이 userLocationData에 map을 돌려서 나온 새로운 배열이라서 어떤 역할을 하는 함수인지 좀 헷갈리네요.

altmit replied:
음 단어가 적절하지는 않은 것 같네요

기존 본인의 동네와 상품에서 적용된 동네가 다를 수 있으니 서로 맞게 보여주기 위함 이니 matching 같은 단어가 더 적절할 수 있겠네요

silvertae commented on src/page/productEditor/ProductEditorPanel.tsx:
```diff
@@ -0,0 +1,457 @@
+    };
+  }, [recommendCategoryData, resetRecommendCategory]);
+
+  useEffect(() => {
+    const changedTitle =
+      editorMode === 'add'
+        ? title.value !== ''
+        : title.value !== productData?.title;
+
+    const handler = setTimeout(() => {
+      if (changedTitle) {
+        refetch();
+      }
+    }, 400);
+
+    return () => {
+      clearTimeout(handler);
+    };
+  }, [editorMode, title.value, productData, refetch]);
```
뭔가 커밋 중에 lodash 라이브러리를 설치하셨던 것 같은데 지금은 지워진 것 같네요.
이 부분에 debounce를 적용하려다가 직접 구현하신건가요?

altmit replied:
obj 깊은 비교 필요할 것 같아서 설치를 했었는데 필요가 없는 것 같아서 삭제 했습니다.

silvertae reviewed (approved):
역대급 작업량이었네요. 어쩐지 제이가 2일 동안 PR을 안 날리실 때부터 알아봤어야 했는데 이렇게 상품 등록/수정 페이지에 많은 기능들이 필요한지 몰랐네요...  작업 분량을 제대로 못 나눈 것 같아 죄송합니다. 😓

최대한 꼼꼼히 보려고 했는데 아무래도 한번에 확인하기 쉽지가 않아 놓친 부분이 좀 있을 것 같네요. 정말 고생 많으셨습니다.

altmit committed:
refactor #82 : editor body에 스크롤 추가

altmit committed:
refactor #82 : 피드백 반영

- resetQueries 불필요한 옵션 삭제
- locationDropdown에서 내 지역설정 삭제
- selectLocationId -> matchLocationId 네이밍 변경

altmit committed:
Merge branch 'dev' into feat/#82-productEditor

altmit committed:
refactor #82 : 경로 변경하지 못한 부분 수정

altmit committed:
feat #82 : 수정 할 데이터 받아오는 api,msw 추가

altmit committed:
refactor #82 : 홈, 디테일 페이지에서 editPanel 수정 가능하게 추가

altmit committed:
refactor #82 : 불필요한 중복 삭제

altmit committed:
refactor #82 : 리팩토링

- status 명확한 타입 지정
- 불필요한 삼항연산자 삭제
- 불필요한 반복 삭제

silvertae commented on src/page/ItemDetails.tsx:
```diff
@@ -15,13 +16,16 @@ import { Dropdown } from '../components/dropdown/Dropdown';
+import { useProductEditorStore } from '../stores/useProductEditorStore';
 import { getElapsedSince } from '../utils/getElapsedSince';
 
+type DetailsStatus = '판매중' | '예약중' | '판매완료';
```
별도 타입으로 분리하니까 더 깔끔하네요 👍

silvertae commented on src/components/ProductItem.tsx:
```diff
@@ -71,7 +89,7 @@ export function ProductItem({
       <Thumbnail src={thumbnailUrl} />
       <Information>
-        <Title>
+        <Title onMouseOver={hoverToFetch}>
```
오 아예 ProductItem의 타이틀에 마우스를 올렸을 때 미리 수정페이지를 열 때 사용할 데이터를 서버에서 받아오고 있네요.
이렇게 하면 판매자 입장에서는 확실히 openPanel 이전에 data를 받아와야 했던 문제랑 빨리 렌더링 해줄 수 있을것 같긴한데 구매자가 홈화면을 볼 때는 좀 달라질 것 같네요.

홈화면에서 스크롤을 내리면서 사용자가 자기가 등록한 아이템이 아니어서 수정이 불가능한데도 마우스를 올리게 되면 실제로 사용하지 않을 데이터에 대해서 요청이 계속 가게되는게 좀 이상하네요. 아예 더보기 버튼의 '게시글 수정' 드롭다운 메뉴에 호버했을 때 fetch 해오는게 더 적절할 것 같습니다.

altmit replied:
저도 그렇게 하고 싶었는데 그럼 공용 컴포넌트를 수정해야 해서 그 부분 보다는 필요한 곳에 맞는 element에 직접 넣는게 맞다고 생각해서 title에 넣었는데 판매자가 아닌 경우를 생각하지 못했네요.

```ts
  const hoverToFetch = () => {
    if (!data && !isError && isSeller) {
      refetch();
    }
  };
```

함수에 조건문을 판매자인 경우만 실행되게 하면 될 것 같네요

altmit replied:
위 방법도 있지만 다른 방법으로 수정해 봤습니다~

silvertae replied:
드롭다운을 여는 더보기 버튼은 어차피 판매자만 볼 수 있으니까 이렇게 하먄 해결이군여 🚀

silvertae commented on src/page/ItemDetails.tsx:
```diff
@@ -134,29 +163,21 @@ export function ItemDetails() {
-                >
-                  판매완료
-                </MenuItem>
+                {itemDetailsData.status.map((state, index) => {
+                  return (
+                    <MenuItem
+                      key={index}
+                      isSelected={state.isSelected}
+                      onAction={() => editStatus(state.name)}
+                    >
+                      {state.name}
+                    </MenuItem>
+                  );
+                })}
```
이 부분은 제가 리팩터링 했어야 했는데 감사합니다. map을 돌리니 코드가 더 짧아지네요.

silvertae commented on src/page/ItemDetails.tsx:
```diff
@@ -180,25 +201,26 @@ export function ItemDetails() {
+          <IconButton
+            styledType="text"
+            onClick={toggleFavorites}
+            onMouseOver={hoverToFetch}
```
관심(하트) 버튼에 hoverToFetch가 붙어있는 건 좀 이상하네요. 차라리 위에서 말했듯 더보기 버튼을 눌렀을 때 열리는 드롭다운에서 게시글 수정 버튼에 있는게 더 좋을것 같다고 생각합니다.

altmit replied:
엥? 이건 왜 여기 붙어있죠 리팩토링하면서 잘 못 붙여넣기 된 것 같네요

altmit committed:
refactor #82 : 판매자가 아닌 경우 호출될 수 있던 로직 수정
